### PR TITLE
Refactor: Track platform tip in Orders table

### PR DESCRIPTION
**Pull request by kewitz**

### CODE REVIEW

**Review by znarf — COMMENTED** [review]
Looks good.

No migration of past entries? Are we gonna support isFeesOnTop / platformFee forever?

While we're at it, would be interesting to track platform tip eligibility as a column in the Orders table.

Both could be done in follow up issues but let's make sure to track them if we go this way.

**Comment by kewitz** [comment]
@znarf I think we can migrate past orders, no problem. And if we want to track platform tip eligibility I'm also doing it now.

File: migrations/20220105170000-add-platformTip-to-orders.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    await queryInterface.addColumn('Orders', 'platformTipAmount', { type: Sequelize.INTEGER, defaultValue: 0 });
+    await queryInterface.addColumn('OrderHistories', 'platformTipAmount', { type: Sequelize.INTEGER, defaultValue: 0 });
+    await queryInterface.addColumn('Orders', 'platformTipEligible', { type: Sequelize.BOOLEAN, defaultValue: false });
+    await queryInterface.addColumn('OrderHistories', 'platformTipEligible', {
+      type: Sequelize.BOOLEAN,
+      defaultValue: false,
```
Comment (znarf) [inline migrations/20220105170000-add-platformTip-to-orders.js:10]: I would default to `NULL`, it's problematic to mark it `FALSE` when we actually don't know or if it's unclear.

File: migrations/20220105170001-migrate-past-platformTip-orders.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+module.exports = {
+  up: async queryInterface => {
+    await queryInterface.sequelize.query(
+      `
+        UPDATE
+          "Orders"
+        SET
+          "platformTipAmount" = ("data"->'platformFee')::int,
+          "platformTipEligible" = true
+        WHERE "data"->>'isFeesOnTop' = 'true';
```
Comment (znarf) [inline migrations/20220105170001-migrate-past-platformTip-orders.js:12]: That's not true, that will be confusing.

File: server/lib/payments.js [hunk 3]
```diff
@@ -832,6 +833,10 @@ export const getHostFee = async (order, host = null) => {
 };
 
 export const isPlatformTipEligible = async (order, host = null) => {
+  if (order.collective?.data?.disablePlatformTips) {
```
Comment (znarf) [inline server/lib/payments.js:836]: What is this?

Comment (kewitz) [reply server/lib/payments.js:836]: This is the edge-case I added for BRL contributions related to Stripe's `aplication_fee` limitation.

Comment (znarf) [reply server/lib/payments.js:836]: Can we remove this now that we're landing https://github.com/opencollective/opencollective-api/pull/7020?

File: migrations/20220105170001-migrate-past-platformTip-orders.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+module.exports = {
+  up: async queryInterface => {
+    await queryInterface.sequelize.query(
+      `
+        UPDATE
+          "Orders"
+        SET
+          "platformTipAmount" = ("data"->'platformFee')::int,
+          "platformTipEligible" = true
```
Comment (znarf) [inline migrations/20220105170001-migrate-past-platformTip-orders.js:11]: Sure, platformTipEligible is TRUE if a platformTip is set. Would be interesting to populate it when there is no platformTip.

Why not try to do it like suggested in https://github.com/opencollective/opencollective-api/pull/7019#discussion_r780403540 ?

File: migrations/20220105170001-migrate-past-platformTip-orders.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        WHERE "data"->>'isFeesOnTop' = 'true' AND ("data"->'platformFee')::int > 0;
+      `,
+    );
+
+    await queryInterface.sequelize.query(
+      `
+        UPDATE
+          "Orders" as o
+        SET
+          "platformTipAmount" = (o."data"->'platformFee')::int,
```
Comment (znarf) [inline migrations/20220105170001-migrate-past-platformTip-orders.js:21]: What if platformFee is not set? OrderId=511771 for example.

File: migrations/20220105170001-migrate-past-platformTip-orders.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+          "platformTipEligible" = true
+        FROM
+          "Transactions" as t
+        WHERE
+          t."OrderId" = o."id"
+          AND t."kind" = 'CONTRIBUTION'
+          AND t."type" = 'CREDIT'
+          AND t."data"->>'platformTipEligible' = 'true'
```
[the remark this answers is not in the document]
Comment (kewitz) [reply migrations/20220105170001-migrate-past-platformTip-orders.js:29]: I don't think I follow what you mean.
Doesn't `platformTipEligible = false` implies that there's no platform tip attached to the order?

File: migrations/20220105170001-migrate-past-platformTip-orders.js (hunk 5 above)
Comment (znarf) [reply migrations/20220105170001-migrate-past-platformTip-orders.js:21]: Is `(o."data"->'platformFee')::int` coalesced to `null` ?

File: migrations/20220105170001-migrate-past-platformTip-orders.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (znarf) [reply migrations/20220105170001-migrate-past-platformTip-orders.js]: Yes, but it's interesting to know that the Contribution was NOT eligible to platform tips with `platformTipEligible=false`.

File: migrations/20220105170001-migrate-past-platformTip-orders.js (hunk 5 above)
Comment (kewitz) [reply migrations/20220105170001-migrate-past-platformTip-orders.js:21]: Exactly. This is valid:
```sql
SELECT NULL::int
```

**Pull request merged**